utils/pub-helpers.rkt

```diff
@@ -0,0 +1,135 @@
+ ;; the publications on that DBLP page we are able to read (currently supports article, book, and
+ ;; inproceedings types).
+ scrape-dblp
+
+ ;; (Listof publication) output-port -> void
+ ;;
+ ;; Writes the given publications to the given port in a format that can easily be copied and pasted
+ ;; into publications.rkt
+ write-pubs)
+
+;; ---------------------------------------------------------------------------------------------------
+
+(require xml
+         net/http-client
+         "../publication-data.rkt")
+
+(define (write-pubs pubs port)
+  (for ([pub pubs])
+  (match-define (publication title authors venue year link) pub)
+  (fprintf port
+"(publication ~s
+             ~s
+             ~s
+             ~s
+             ~s)\n"
+           title authors venue year link)))
+
+;; ---------------------------------------------------------------------------------------------------
+;; DBLP
+
+(define supported-record-types `(article book inproceedings))
+
+(define (scrape-dblp uri)
+  (define-values (status headers xml-stream) (http-sendrecv "dblp.uni-trier.de" uri))
+  (define pub-records
+    (map (compose1 first element-content)
+         (get-children-of-name (document-element (read-xml xml-stream)) 'r)))
+  (for/list ([record pub-records] #:when (member (record-type record) supported-record-types))
+    (dblp-record->pub record)))
+
+(define (dblp-record->pub record)
+  (define authors
+    (map (compose1 pcdata-string first element-content) (get-children-of-name record 'author)))
+  (define author-text
+    (match (length authors)
+      [1 (first authors)]
+      [2 (string-append (first authors) " and " (second authors))]
+      [_ (string-join authors ", " #:before-last ", and ")]))
+  (define venue
+    (match (element-name record)
+      ['article (element-text (get-child-of-name record 'journal))]
+      ['book (element-text (get-child-of-name record 'publisher))]
+      ['inproceedings
+       (let ([booktitle (element-text (get-child-of-name record 'booktitle))])
+         (hash-ref booktitle-venue-map booktitle booktitle))]
+      [other-type (error 'record->pub "Received unsupported record type ~s" other-type)]))
+  (define url
+    (match (get-children-of-name record 'ee)
+      [(list url) (element-text url)]
+      [_ #f]))
+  (publication (element-text (get-child-of-name record 'title))
+               author-text
+               venue
+               (string->number (element-text (get-child-of-name record 'year)))
+               url))
+
+(define booktitle-venue-map
+  (hash
+   "OOPSLA" "Object-Oriented Programming Systems, Languages, and Applications (OOPSLA)"
+   "PLDI" "Programming Language Design and Implementation (PLDI)"
+   "ICFP" "International Conference on Functional Programming (ICFP)"
+   "POPL" "Principles of Programming Languages (POPL)"
+   "ESOP" "European Symposium on Programming (ESOP)"
+   "JFP" "Journal of Functional Programming (JFP)"
```

Comment:
Wait now I'm confused... where did the "J. Funct. Program" abbreviation come from?